Quicker build

pdmosses opened the pull request:
This PR uses the cached site-nav to determine the position of each page in the navigation hierarchy. [Profiling](https://just-the-docs.github.io/just-the-docs-tests/tests/profiles/index/) shows that it significantly reduces the v0.7.0 build time for larger sites – especially with Jekyll 3. The improvement is due mainly to the elimination of loops and filters that depended on the entire site. 

For example:

- [endoflife.date](https://just-the-docs.github.io/just-the-docs-tests/tests/profiles/endoflife.date/index/) drops from 32s to 11s
- [machinetranslate.org](https://just-the-docs.github.io/just-the-docs-tests/tests/profiles/machinetranslate.org/index/) drops from 267s to ~~73s~~ 56s

### Testing

[Just the Docs Tests](https://just-the-docs.github.io/just-the-docs-tests/) is currently built using this PR branch. With JS disabled, the highlighting and unfolding of the navigation panel should be as with v0.7.0.

A diff of the tests site built with this PR branch compared to v0.7.0 should report no significant changes. For more rigorous testing, diffs of variants of the tests site should also be inspected.

pdmosses commented:
Diffing the `just-the-docs-tests` site between this PR and v0.7.0 revealed two (minor) bugs: one concerning empty collections, the other regarding pages that are implicitly excluded because of the exclusion of their parents or grandparents. They have both been fixed by the above commit.

The following `diff` command suppresses reports of irrelevant differences:

```sh
diff -rw -x "*.js*" -x "*.map" -I "published_time" -I "dateModified" -I "<meta property=" -I "<link rel=" -I "@context" _site _site-0.7.0
```

Regarding the options used:

* `-r`: application of diff recursively to common subdirectories encountered.
* `-w`: ignore whitespace (blanks and tabs).
* `-x "*.js*" -x "*.map"`: ignore JavaScript and CSS map files and subdirectories.
* `-I "published_time" -I "dateModified" ...`: ignore changes of build-dependent metadata.

pdmosses commented:
Diffing with `just-the-docs-tests` revealed that the previous WIP commit to `components/breadcrumbs.html` didn't generate the correct breadcrumbs for pages that are implicitly excluded because of the exclusion of their parents or grandparents. The above commit fixes that.

I've not been able to find a more efficient way of generating breadcrumbs for (implicitly or explicitly) excluded pages with Jekyll 3. The version of the machinetranslate.org site used for profiling has 49 excluded pages, and it takes about 14s for Jekyll 3 to generate their breadcrumbs; generating the breadcrumbs for the more than 500 non-excluded pages takes about 4s. With Jekyll 4, the time for generating the breadcrumbs for _all_ pages is less than 3s!

mattxwang commented:
Ah! This is on my radar, I've just been bogged down with lots of work :/ will do my best to review this week!

mattxwang commented on _includes/components/breadcrumbs.html:
```diff
@@ -8,8 +8,88 @@
+  {%- if nav_split_next contains "</ul>" -%}
+    {%- assign nav_endless = nav_split_next | remove: "</ul>" -%}
+    {%- assign nav_ends =
+          nav_split_next.size | minus: nav_endless.size | divided_by: 5 -%}
```
Quick question: why is the `divided_by: 5` filter here? I'm not sure that I understand what `nav_ends` is supposed to be.

pdmosses replied:
Indeed, that should have been explained. Is it adequate to add the following in the preceding comment?

```
    To compute the number of "</ul>" strings in nav_split_next, removing them all 
    gives the string nav_endless. The difference between the sizes of 
    nav_split_next and nav_endless is the number of removed strings times 5.
```

A corresponding comment in `_includes/css/activation.scss.liquid` could be:

```
  To compute the number of nav_list strings in a nav_split, removing them all 
  gives the string nav_listless. The difference between the sizes of nav_split 
  and nav_listless is the number of removed strings times their size.
  Similarly to compute nav_ends, where the size of the removed strings is 5.
```

mattxwang commented on _includes/css/activation.scss.liquid:
```diff
@@ -37,89 +34,108 @@
+    {%- assign nav_endless = nav_split | remove: "</ul>" -%}
+    {%- assign nav_ends = nav_split.size | minus: nav_endless.size | divided_by: 5 -%}
+    
+    {%- if nav_lists >= 2 or nav_lists >= 1 and forloop.first or forloop.last -%}
```
Suggestion: could you add a comment here that explains what the conditions are checking for?

pdmosses replied:
I'll try – but the case analysis that underlies the current code is unfortunately quite complicated…

mattxwang reviewed (commented):
Thanks for your patience! Finally had the time to sit down and really look at this PR. Great job - especially with some of the blanket optimizations for all pages (ex removing nested checks for `page.parent`)! I think this'll be a well-received PR, and happy to merge it in shortly. I will say that I probably couldn't reproduce this code myself (which may lead to maintenance implications down the line).

I also reproduced your `diff` locally using `just-the-docs-test`. I will say, I didn't find any differences between the two directories even after omitting each of the flags (i.e. running `diff -r _site _site-0.7.0`), though after inserting my own arbitrary changes (randomly editing files) I don't think this is a fluke. Overall, that's very promising!

One broad concern I have is this does tie the breadcrumbs component quite tightly to the site navigation; if users override the structure (ex by adding another class to the elements with `nav-list-link`, `nav-category`, `nav-list`, etc.). I don't think this means we shouldn't implement this feature (as the performance speedup is astounding!), but perhaps we should document this somewhere and/or allow an escape hatch. What are your thoughts? I think just documenting it is fine, though adding an escape hatch is pretty trivial (just adding an extra `site.` top-level condition to default to the `nav_excludes: true` behaviour).

pdmosses commented:
> I will say that I probably couldn't reproduce this code myself (which may lead to maintenance implicants down the line).

I must admit that it's at the limit of what I can implement in Liquid/Jekyll...

Apart from the usual problem with the clumsiness of Liquid programming (such as the roundabout way of calculating the number of occurrences of a substring) an aggravating factor is the non-uniformity of our navigation HTML regarding collections:

- The structure of foldable collections has an extra level of nav-list.
- The possibility of empty collections gives rise to any number of empty nav-list elements between two nav-list-items.

In fact I've just discovered a bug in the current activation code: if an _empty_ collection has `nav_fold: true`, the computed index of the following collection is wrong! ☹️ For example, make **Collection 0** foldable, disable JS, and click on the top page in **Collection 1**: the top page in **Collection 2** is highlighted... I'll obviously need to fix that.[^1] I could also try to simplify/clarify the existing code a bit in the process.

[^1]: Empty collections aren't at all useful, but they could easily arise while developing a website, as it's natural to configure collections before adding pages to them. The resulting navigation would be broken when browsing such an interim website with JS disabled.

pdmosses commented:
> I also reproduced your `diff` locally using `just-the-docs-test`. I will say, I didn't find any differences between the two directories even after omitting each of the flags (i.e. running `diff -r _site _site-0.7.0`), though after inserting my own arbitrary changes (randomly editing files) I don't think this is a fluke.

There should _always_ be a difference on the home page, because it displays whether the site was built using the v0.7.0 release or the PR branch (see https://just-the-docs.github.io/just-the-docs-tests/).

When I generate `_site-0.7.0` I comment-out the `github` and `ref` parameters in `Gemfile`, followed by:

```sh
bundle update
bundle lock --add-platform x86_64-linux
bundle exec Jekyll build
mv _site _site-0.7.0
```
After reverting the `Gemfile` change and rebuilding `_site`, running `diff -r _site _site-0.7.0` generates a _lot_ of output for me.

pdmosses commented:
> I'll obviously need to fix that. 

Done.

> I could also try to simplify/clarify the existing code a bit in the process.

I've now separated the effects for the various cases. I've also added a semi-formal specification of the assumptions, and clarified the explanations.

I hope that this has made the code significantly easier to follow, and to review…

I've also added some extra configuration files to `just-the-docs-tests` to facilitate testing with different patterns of elements in `site_nav`.

didacrios reviewed (approved):
Thanks!

mattxwang commented on _includes/css/activation.scss.liquid:
```diff
@@ -236,9 +288,9 @@
-{%- if activation_third_level_index %} > ul > li:nth-child({{ activation_third_level_index }})
+{{ activation_collection_prefix }} > li:nth-child({{ nav_levels[1] }})
+{%- if nav_levels[2] %} > ul > li:nth-child({{ nav_levels[2] }})
+{%- if nav_levels[3] %} > ul > li:nth-child({{ nav_levels[3] }})
```
I *think* this line also has the wrong `nav_levels[3]`? Which would explain

```
diff --color -rw -x *.js* -x *.map -I published_time -I dateModified -I <meta property= -I <link rel= -I @context _site/navigation/order/mixture/0/index.html _site-0.7.0/navigation/order/mixture/0/index.html
19c19
< .site-nav > ul:nth-of-type(8) > li > ul > li:nth-child(6) > ul > li:nth-child(5) > ul > li:nth-child(3) > a {
---
> .site-nav > ul:nth-of-type(8) > li > ul > li:nth-child(6) > ul > li:nth-child(5) > ul > li:nth-child(4) > a {
```

and

```
diff --color -rw -x *.js* -x *.map -I published_time -I dateModified -I <meta property= -I <link rel= -I @context _site/navigation/order/mixture/0.0/index.html _site-0.7.0/navigation/order/mixture/0.0/index.html
19c19
< .site-nav > ul:nth-of-type(8) > li > ul > li:nth-child(6) > ul > li:nth-child(5) > ul > li:nth-child(4) > a {
---
> .site-nav > ul:nth-of-type(8) > li > ul > li:nth-child(6) > ul > li:nth-child(5) > ul > li:nth-child(3) > a {
```

pdmosses replied:
I can't imagine any way in which the changes in this PR could make a difference to the order of the cached navigation panel.

The diff suggested that Liquid/Jekyll doesn't sort `0` and `0.0` in a particular order, and I've now verified that.

mattxwang reviewed (commented):
So sorry for the late response on this - work has really been slamming me :( 

I re-ran your provided `diff` command against `main` with just-the-docs-tests, and I did see a discrepancy in the generated CSS:

```
$ diff -rw -x "*.js*" -x "*.map" -I "published_time" -I "dateModified" -I "<meta property=" -I "<link rel=" -I "@context" _site _site-0.7.0
[OMITTED]
---
>     <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>remote: https://rubygems.org/
diff --color -rw -x *.js* -x *.map -I published_time -I dateModified -I <meta property= -I <link rel= -I @context _site/navigation/order/mixture/0/index.html _site-0.7.0/navigation/order/mixture/0/index.html
12c12
< .site-nav > ul:nth-of-type(8) > li > ul > li > ul > li > ul > li:not(:nth-child(3)) > a {
---
> .site-nav > ul:nth-of-type(8) > li > ul > li > ul > li > ul > li:not(:nth-child(4)) > a {
19c19
< .site-nav > ul:nth-of-type(8) > li > ul > li:nth-child(6) > ul > li:nth-child(5) > ul > li:nth-child(3) > a {
---
> .site-nav > ul:nth-of-type(8) > li > ul > li:nth-child(6) > ul > li:nth-child(5) > ul > li:nth-child(4) > a {
diff --color -rw -x *.js* -x *.map -I published_time -I dateModified -I <meta property= -I <link rel= -I @context _site/navigation/order/mixture/0.0/index.html _site-0.7.0/navigation/order/mixture/0.0/index.html
12c12
< .site-nav > ul:nth-of-type(8) > li > ul > li > ul > li > ul > li:not(:nth-child(4)) > a {
---
> .site-nav > ul:nth-of-type(8) > li > ul > li > ul > li > ul > li:not(:nth-child(3)) > a {
19c19
< .site-nav > ul:nth-of-type(8) > li > ul > li:nth-child(6) > ul > li:nth-child(5) > ul > li:nth-child(4) > a {
---
> .site-nav > ul:nth-of-type(8) > li > ul > li:nth-child(6) > ul > li:nth-child(5) > ul > li:nth-child(3) > a {
```

(the same occurs when diffing w/ `main`)

I've threaded a comment for the suspect line, which seems to be an incorrect `nav_levels[3]` in an edge case of nav items `0` and `0.0` within the mixed navigation order collection. What are your thoughts?

(also overall, recognizing that I've been super behind on my JtD maintenance obligations. Will try to merge this in and push out 0.8.0 asap!)

pdmosses commented on _includes/css/activation.scss.liquid:
```diff
@@ -197,25 +249,25 @@
 {{ activation_collection_prefix }} > li > a,
 {{ activation_collection_prefix }} > li > ul > li > a,
-{{ activation_collection_prefix }} > li > ul > li > ul > li:not(:nth-child({{ activation_third_level_index }})) > a {
+{{ activation_collection_prefix }} > li > ul > li > ul > li:not(:nth-child({{ nav_levels[3] }})) > a {
```
@mattxwang thanks for finding time for re-reviewing this PR – especially as it's probably even harder to review than it was to implement…

In fact the varying sort order of the `0` and `0.0` pages was already [mentioned](https://github.com/just-the-docs/just-the-docs/pull/1244#issuecomment-1656819401) when testing #1244, which was merged in v0.6.0. So it's definitely not caused by the present PR.

Try running the following code:

```liquid
{% assign a = "" | split: "" | push: 0 | push: 0.0 | sort %}
{% assign b = "" | split: "" | push: 0.0 | push: 0 | sort %}

- {{ a | inspect }}
- {{ b | inspect }}
```

Testing locally, that produces:

![image](https://github.com/just-the-docs/just-the-docs/assets/18308236/6ff1a9af-fded-4c5d-bded-18f1f289901a)

I suspect that the order of the arrays of pages accessed via `site` variables is (at least to some extent) random. Since sorting doesn't determine a specific order for `0` and `0.0`, the nav links to pages with those values can appear in either order.

I should probably remove one of the pages, and add a warning about mixing integers and floats. Then the false alarm in the diff will definitely disappear! 😃

mattxwang replied:
Oh! Wow, that completely slipped my mind - it's been a while since I reviewed #1244. Thank you for the patient explanation! Apologies for leading you down an incorrect rabbit hole re: `nav_levels`.

> I suspect that the order of the arrays of pages accessed via site variables is (at least to some extent) random. Since sorting doesn't determine a specific order for 0 and 0.0, the nav links to pages with those values can appear in either order.

Shopify's [`sort`](https://shopify.github.io/liquid/filters/sort/) documentation doesn't mention anything about stable sorting, but my (uninformed) guess is that `0` and `0.0` are both just [`Numbers`](https://shopify.github.io/liquid/basics/types/#number) and must evaluate to the same underlying number (which might cause this behaviour - your example is very helpful here). I'm realizing now that I have no concrete idea of how Liquid's type/value system really works, when I probably should!

About to hop into a few meetings, but I'll take one last look and likely approve; the comments you added in `activation.scss.liquid` (w/ the CFG) were quite helpful. I might also look through the Liquid source code and see if I can get any hints as to why this behaviour might happen, but that shouldn't block this PR.

mattxwang reviewed (approved):
Okay, have given this one more glance + have run on endoflife.date (`main` against this branch) with:

```
$ diff -rw -x "**/*.js*" -x "*.map" -I "published_time" -I "dateModified" -I "<meta property=" -I "<link rel=" -I "@context" -I "DTSTAMP" -I "UID" _site _site-main
```

and don't see any red flags (added the `DTSTAMP` and `UID` to ignore generation artifacts). I think this LGTM!

---

I did a quick check down the rabbit hole, and:

- there are [a handful of GitHub issues about Liquid's sort function](https://github.com/Shopify/liquid/issues?q=is%3Aissue+sort), including a request on documenting sort stability (https://github.com/Shopify/liquid/issues/897) and on numeric sort (https://github.com/Shopify/liquid/issues/980). None demonstrate this exact problem though. Other commenters make claims about how Liquid's sort works (e.g. converts input -> number -> string), but not sure how verifiable this is.
- did a teensy bit of digging, and Liquid's [sort filter is defined as a standard filter](https://github.com/Shopify/liquid/blob/4a4fe3c72a91cc8a26869ed2dad8e3fcac33565a/lib/liquid/standardfilters.rb#L370). The tl;dr is that it converts the input to a Liquid [`InputIterator`](https://github.com/Shopify/liquid/blob/4a4fe3c72a91cc8a26869ed2dad8e3fcac33565a/lib/liquid/standardfilters.rb#L953) and then uses the inherited sort method from the Ruby core's [`Enumerable` module](https://ruby-doc.org/3.2.2/Enumerable.html). [Enumerable's `sort` docs](https://ruby-doc.org/3.2.2/Enumerable.html#method-i-sort) say:

> The ordering of equal elements is indeterminate and may be unstable.

So I think this is where the rabbit hole stops for now. It's good to know though! I am slightly surprised that there's not more on this (given that I could see this having very real business impacts on Shopify's customers).
